rust_porting_cookbook.md

```diff
@@ -0,0 +1,240 @@
+The easy work around is to comment out `lisp_fn` until the compile succeeds
+then enable it to do a final build and begin testing.
+
+The elisp name of the function is derived from the Rust name, with
+underscores replaced by hyphens.  If that is not possible (like for
+the function `+`), you can give an elisp name as an argument to
+`lisp_fn`, like `#[lisp_fn(name = "+")]`.
+
+Optional arguments are also possible: to make the minimum number of
+arguments from elisp different from the number of Rust arguments,
+pass a `min = "n"` argument.
+
+The docstring of the function should be the same as the docstring
+in the C code.  (Don't wonder about it being a comment there, Emacs
+has some magic that extracts it into a separate file.)
+
+Finally, delete the old C definition.
+
+You're done! Compile Remacs, try your function with `M-x ielm`, and
+open a pull request. Fame and glory await!
+
+## Porting Widely Used C Functions
+
+If a Rust function replaces a C function that is used elsewhere in
+the C codebase, it needs to be exported. If the function is not a Lisp
+function (i.e. doesn't use the `#[lisp_fn]` macro), you need to
+manually mark it as `#[no_mangle]` and `extern "C"` to be exported
+with the correct ABI. The build scripts handle the export automatically
+however you need to maintain an entry in a C header to keep the compilation
+happy.
+
+## Source code style guide
+
+In order to pass Travis checks on pull requests, the source has to
+be formatted according to the default style of `rustfmt-nightly`, version 0.2.9.
+To do that, install `rustfmt`:
+
+```
+$ cargo install rustfmt-nightly
+```
+
+Please note that this is not the old `-nightly` version of the
+`rustfmt` crate, 0.9.0. It is a different crate with a lower version
+number which is in fact newer. See
+https://users.rust-lang.org/t/rustfmt-releases/11357 and the linked
+blog post for details. In short, `rustfmt` is being written with a new
+API which only supports nightly at the moment and the split crate
+situation will be resolved eventually.
+
+Then you can run this in the checkout root to reformat all Rust code:
+
+```
+$ make rustfmt
+```
+
+## Running tests
+
+Run elisp and Rust tests in toplevel directory. If run in a subdirectory,
+only run the tests in that directory.
+
+* `make check`
+  Run all tests as defined in the directory. Expensive tests are
+  suppressed. The result of the tests for <filename>.el is stored in
+  <filename>.log.
+
+* `make check-maybe`
+  Like "make check", but run only the tests for files that have been
+  modified since the last build.
+
+## Rust Porting Tips
+
+### C Functions
+
+When writing a Rust version of a C function, give it the same name and
+same arguments. If this isn't appropriate, docstrings should say the
+equivalent C function to help future porters.
+
+For example, `make_natnum` mentions that it can be used
+in place of `XSETFASTINT`.
+
+### C Macros
+
+For C macros, we try to define a fairly equivalent Rust
```

Comment:
a single macro/group of macros in C are often replaced by rust macro. eg: xsignal, call, wrong_type_arg etc..

Reply:
Agreed.

We were talking in Gitter channel last night about how this doc might work better as a Wiki page.

Reply:
Yeah. That would be better